pkg/network/protocols/events/batch_reader.go

```diff
@@ -0,0 +1,114 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf
+// +build linux_bpf
+
+package events
+
+import (
+	"sync"
+	"unsafe"
+
+	"github.com/cilium/ebpf"
+)
+
+var batchPool = sync.Pool{
+	New: func() interface{} {
+		return new(batch)
+	},
+}
+
+type batchReader struct {
+	sync.Mutex
+	numCPUs    int
+	batchMap   *ebpf.Map
+	offsets    *offsetManager
+	workerPool *workerPool
+	stopped    bool
+}
+
+func newBatchReader(offsetManager *offsetManager, batchMap *ebpf.Map, numCPUs int) (*batchReader, error) {
+	// initialize eBPF maps
+	batch := new(batch)
+	for i := 0; i < numCPUs; i++ {
+		for j := 0; j < batchPagesPerCPU; j++ {
+			key := &batchKey{Cpu: uint32(i), Num: uint32(j)}
+			err := batchMap.Put(unsafe.Pointer(key), unsafe.Pointer(batch))
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return &batchReader{
+		numCPUs:    numCPUs,
+		offsets:    offsetManager,
+		batchMap:   batchMap,
+		workerPool: newWorkerPool(20),
+	}, nil
+}
+
+// ReadAll batches from eBPF (concurrently) and execute the given
+// callback function for each
+func (r *batchReader) ReadAll(f func(cpu int, b *batch)) {
+	// This lock is used only used for the purposes of synchronizing termination
+	// and it's only held while *enqueing* the jobs.
+	r.Lock()
+	if r.stopped {
+		r.Unlock()
+		return
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(r.numCPUs)
+
+	for i := 0; i < r.numCPUs; i++ {
+		cpu := i // required to properly capture this variable in the function closure
+		r.workerPool.Do(func() {
+			defer wg.Done()
+			batchID, key := r.generateBatchKey(cpu)
+
+			b := batchPool.Get().(*batch)
+			defer func() {
+				*b = batch{}
```

Comment:
doesn't you allocate the object and feed the pool in that operation? instead of using the pool's allocator whenever needed, we are allocating the objects

Reply:
this is only zeroing fields from an existing struct (`*b`). there is no allocation involved